### PR TITLE
feat(mr): add cluster roles for model registry administration, fixes RHOAIENG-5112

### PR DESCRIPTION
**Pull request by dhirajsb**

### CODE REVIEW

**Comment by dhirajsb** [comment]
@zdtsw @VaishnaviHire can we get this change in for model registry operator for odh 2.10.1?

File: components/modelregistry/modelregistry.go [hunk 1]
```diff
@@ -82,6 +85,44 @@ func (m *ModelRegistry) ReconcileComponent(_ context.Context, cli client.Client,
 				return fmt.Errorf("failed to update image from %s : %w", Path, err)
 			}
 		}
+
+		// Create model registry viewer and editor ClusterRoles if not present
```
Comment (zdtsw) [inline components/modelregistry/modelregistry.go:89]: any reason not add the clusterrole in model registery part either as manifests or logic, instead of having it here?

File: pkg/metadata/labels/types.go [hunk 2]
```diff
@@ -15,9 +15,11 @@ var K8SCommon = struct {
 var ODH = struct {
 	OwnedNamespace string
 	Component      func(string) string
+	ComponentName  string
 }{
 	OwnedNamespace: "opendatahub.io/generated-namespace",
 	Component: func(name string) string {
 		return ODHAppPrefix + "/" + name
 	},
+	ComponentName: "component.opendatahub.io/name",
```
Comment (bartoszmajsak) [inline pkg/metadata/labels/types.go:24]: Is this label already established (i.e., does anything currently depend on it?), or is it a new addition we're considering?

I also see a need for better means of defining taxonomy for ODH. It seems your label somewhat overlaps with `platform.opendatahub.io/component` I put in [my proposal](https://github.com/opendatahub-io/architecture-decision-records/pull/26).

If it's a new addition, perhaps moving forward with [the proposal](https://github.com/opendatahub-io/architecture-decision-records/pull/26) together would be beneficial. 

I think it would be good first to have it fleshed out and agreed on, so we can proceed with enriching all components similarly.

WDYT @dhirajsb?

Comment (bartoszmajsak) [reply pkg/metadata/labels/types.go:24]: > [!NOTE]
> This file is based on existing (legacy?) labels. The intention was to gather them together so they're easy to discover - a bit more explanation in the related PR #941

File: components/modelregistry/modelregistry.go (hunk 1 above)
Comment (bartoszmajsak) [reply components/modelregistry/modelregistry.go:89]: There is nothing dynamic or dependent on the DSC/DSCI config in this so we could keep it under kustomize umbrella. I agree this fits better in `overlays/odh` if it's ODH-specific.

**Review by bartoszmajsak — CHANGES_REQUESTED** [review]

File: pkg/metadata/labels/types.go (hunk 2 above)
Comment (dhirajsb) [reply pkg/metadata/labels/types.go:24]: I'm happy with moving to a newer convention if there is one.

File: components/modelregistry/modelregistry.go (hunk 1 above)
Comment (dhirajsb) [reply components/modelregistry/modelregistry.go:89]: Basically, these roles have the same lifecycle requirements for these clusterroles as the shared namespace in #930.

File: pkg/metadata/labels/types.go (hunk 2 above)
Comment (dhirajsb) [reply pkg/metadata/labels/types.go:24]: But, my understanding is that it would have to be a big refactor across all components to standardize this new label.

Comment (bartoszmajsak) [reply pkg/metadata/labels/types.go:24]: I might be wrong, but where is `component.opendatahub.io/name` currently in use?

File: components/modelregistry/modelregistry.go (hunk 1 above)
Comment (bartoszmajsak) [reply components/modelregistry/modelregistry.go:89]: Let's blame it for lack of coffee :) It was early for me. Maybe worth adding a comment around this piece as of why it's here, not in the manifest. So some other lost soul like me wouldn't waste ppls time asking questions.

Comment (bartoszmajsak) [reply components/modelregistry/modelregistry.go:89]: But I wonder - couldn't Model Registry operator take care of the needed roles instead?

Comment (dhirajsb) [reply components/modelregistry/modelregistry.go:89]: True, that was an option I considered. But, I figured these were a user facing/odh platform concern and hence better created in the odh operator instead.

File: pkg/metadata/labels/types.go (hunk 2 above)
Comment (dhirajsb) [reply pkg/metadata/labels/types.go:24]: Yeah, that label is being referenced in [multiple places](https://github.com/search?q=org%3Aopendatahub-io%20component.opendatahub.io%2Fname&type=code). I'm guessing the dashboard might also be using that to search for resources belonging to a component.

File: components/modelregistry/modelregistry.go (hunk 1 above)
Comment (zdtsw) [reply components/modelregistry/modelregistry.go:89]: all in all, the reason to creata 2  clusterroles  is that they should not be removed when model registry component is set to Removed, is this correct?
two ways i can think of making this work:
- use manifests from Model Registry +  filter out these 2 in ODH operator to not "delete"
- creating them in ODH operator (what this PR is doing)

Comment (zdtsw) [reply components/modelregistry/modelregistry.go:89]: when i say filter out:
- either we do the hard code filter out like how we treated some resources like dashboard's CRD
- or gather the resources you do not want to delete (i.e these 2 into one decidated folder under kustomize), thenn ODH operator call `DeployManifestsFromPath(...., true)` on this path

File: pkg/cluster/resources.go [hunk 3]
```diff
@@ -142,3 +142,33 @@ func CreateNamespace(cli client.Client, namespace string, metaOptions ...MetaOpt
...
+			if err != nil && !apierrs.IsAlreadyExists(err) {
+				return nil, err
+			}
+			desiredClusterRole.DeepCopyInto(foundClusterRole)
+		} else {
+			return nil, err
+		}
+	}
+
+	return foundClusterRole, nil
```
Comment (bartoszmajsak) [inline pkg/cluster/resources.go:173]: A personal-taste suggestion. It might improve readability a bit as cyclomatic complexity is lower (lazy 🧠 can comprehend quicker)

```suggestion
	if getErr := cli.Get(context.TODO(), client.ObjectKey{Name: name}, foundClusterRole); client.IgnoreNotFound(getErr) != nil {
		return nil, getErr
	}

	if createErr := cli.Create(context.TODO(), desiredClusterRole); client.IgnoreAlreadyExists(createErr) != nil {
		return nil, createErr
	}

	desiredClusterRole.DeepCopyInto(foundClusterRole)
	return foundClusterRole, nil
```

> [!WARNING]
> Suggestion takes into account nit rename of `clusterRole` to `name`. So choose wisely ;)

File: pkg/cluster/resources.go [hunk 4]
```diff
@@ -142,3 +142,33 @@ func CreateNamespace(cli client.Client, namespace string, metaOptions ...MetaOpt
 
 	return foundNamespace, nil
 }
+
+// CreateClusterRole creates cluster role required by model registry component.
```
Comment (bartoszmajsak) [inline pkg/cluster/resources.go:146]: It's a generic function and others might need it to :) Thus a statement about the particular need is not needed, unless we are not allowed to touch it ;)

```suggestion
// CreateClusterRole creates cluster role based on define PolicyRules and optional metadata fields.
```

File: pkg/cluster/resources.go [hunk 5]
```diff
@@ -142,3 +142,33 @@ func CreateNamespace(cli client.Client, namespace string, metaOptions ...MetaOpt
...
+func CreateClusterRole(cli client.Client, clusterRole string, rules []authv1.PolicyRule, metaOptions ...MetaOptions) (*authv1.ClusterRole, error) {
+	desiredClusterRole := &authv1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: clusterRole,
+		},
+		Rules: rules,
+	}
```
Comment (bartoszmajsak) [inline pkg/cluster/resources.go:153]: nit

```suggestion
func CreateClusterRole(cli client.Client, name string, rules []authv1.PolicyRule, metaOptions ...MetaOptions) (*authv1.ClusterRole, error) {
	desiredClusterRole := &authv1.ClusterRole{
		ObjectMeta: metav1.ObjectMeta{
			Name: name,
		},
		Rules: rules,
	}
```

File: components/modelregistry/modelregistry.go [hunk 6]
```diff
@@ -82,6 +85,46 @@ func (m *ModelRegistry) ReconcileComponent(_ context.Context, cli client.Client,
...
+					Verbs: []string{"get"}},
+			},
+			cluster.WithLabels(labels.ODH.Component(ComponentName), "true",
+				labels.ODH.ComponentName, ComponentName,
+				"app.kubernetes.io/name", "clusterrole",
+				"app.kubernetes.io/instance", "modelregistry-editor-role",
+				"app.kubernetes.io/component", "rbac",
+				"app.kubernetes.io/created-by", "model-registry-operator",
+				"app.kubernetes.io/part-of", "model-registry-operator",
+				"app.kubernetes.io/managed-by", "model-registry-operator"))
```
Comment (bartoszmajsak) [inline components/modelregistry/modelregistry.go:124]: Maybe we can add those labels to `pkg/metadata/labels` and re-use? I think this will be used more moving forward, so extracting it to constant might help.

https://github.com/opendatahub-io/opendatahub-operator/blob/08b74a7ce1dd0e005151893159d0f3a8846c3c87/pkg/metadata/labels/types.go#L13-L17

File: pkg/cluster/resources.go
Comment (bartoszmajsak) [inline pkg/cluster/resources.go:1]: We do have some tests for these operations: 

https://github.com/opendatahub-io/opendatahub-operator/blob/08b74a7ce1dd0e005151893159d0f3a8846c3c87/pkg/cluster/cluster_operations_int_test.go

Adding a few edge cases for the new one would be more than awesome.

**Review by bartoszmajsak — COMMENTED** [review]
I am fine with this approach as it's very specific for our use case. I shared a few applicable suggestions for your consideration.

I have two open questions about the functionality, not the implementation as such:

1. Are we fine with "create and do not reconcile"? We deliberately ignore if the roles already exist. Is it because something else can and should be able to alter those roles? 
2. Who is driving role-binding in this case? Is it a dashboard?

In the retrospect, by looking at the number of comments related to WHY this is done the way it is, I think it would be great to add a PR description with the rationale behind the chosen path. That would improve merge lead time.

**Comment by zdtsw** [comment]
want to do a follow up on this PR: @dhirajsb do we still need it?  + comments from https://github.com/opendatahub-io/opendatahub-operator/pull/953#pullrequestreview-2010843475

**Comment by dhirajsb** [comment]
@zdtsw I'm going to refactor this PR to use manifests for both shared namespace and these Roles. I'll push the new version to this PR this week.

**Comment by dhirajsb** [comment]
@zdtsw @bartoszmajsak simplified this PR to move model registry operator resources into an extras manifests directory in model registry operator.

**Comment by dhirajsb** [comment]
@zdtsw the unit test is failing at https://github.com/opendatahub-io/opendatahub-operator/blob/46ef53e243928f820a9335a676d4a42c174d3d22/controllers/dscinitialization/dscinitialization_test.go#L296-L295 which expects no rolebindings at the end of the test. Any ideas, what could be causing that to fail?

**Comment by dhirajsb** [comment]
@VaishnaviHire can we merge this in for the coming odh  2.13 release?

**Comment by zdtsw** [comment]
@dhirajsb sorry, did not see your comments above.
any reason to have a dedicated "extra" folder for these roles? other than that, i am fine with the code

**Comment by dhirajsb** [comment]
@zdtsw the extras folder is for keeping the base resources separate from the extras. The base ones also creates resources in a common namespace, etc. so it allows extras manifests more flexibility if it's needed in the future.

**Review by zdtsw — APPROVED** [review]

**Comment by tarilabs** [comment]
Based on https://github.com/opendatahub-io/opendatahub-operator/pull/953#pullrequestreview-2092121712
should the label `do-not-merge/hold` be removed to allow bot to merge this PR @zdtsw ?
thanks
